File: examples/cs/gmrpl/gmrpl.cs
/*
 * Copyright 2007-2016 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */



/*! \file gmrpl.cs
 *
 * A C# example demonstrating the request/reply functionality.
 * The supplied gmreq_cs will provide the request.
 * gmrpl_cs  should be started before gmreq_cs.
*/


namespace gmrpl
{
	using GMSEC.API;
	using GMSEC.API.UTIL;
	using GMSEC.API.Example.Common;
	using System;
	using System.Diagnostics;
	using System.Collections.Generic;


	class gmrpl : GmsecExample
	{
		private Connection conn;
        private List<SubscriptionInfo> info = new List<SubscriptionInfo>(); 

		public gmrpl()
		{
		}


		private void PrintUsage()
		{
			Console.Out.WriteLine(
				"\nusage: gmrpl_cs connectiontype=<middleware> " +
				"[ <parameter>=<value> ]\n" +
				"\n\tNote that the parameter 'connectiontype' is required. " + 
				"\n\tThe rest of other parameters are optional.\n" +
				"\n\tserver=<server name> " +
				"(required if middleware is not bolt/MB locally)" +
				"\n\tsubject=<subject name>" +
				"\n\tmsg_timeout_ms=<timeout period (milliseconds)>" +
				"\n\tprog_timeout_s=<timeout period (seconds)>" +
				"\n\t\tNote: msg_timeout_ms must be defined to use prog_timeout_s" +
				"\n\tloglevel=<log level>" +
				"\n\tcfgfile=<config_filename.xml>" +
				"\n\nFor more information, see API User's Guide\n");
		}


		private void CleanUp()
		{
			if (conn != null)
			{
                for (int i = info.Count - 1; i >= 0; i--)
                {
                    Log.Info("Unsubscribing to " + info[i].GetSubject());
                    var temp = info[i];
                    conn.Unsubscribe(ref temp);
                    info.RemoveAt(i);
                }
                
                conn.Disconnect();

				Connection.Destroy(ref conn);
			}
		}


		private void Run(string[] args)
		{
			Config config = new Config(args);

			Initialize(config);

			AddToConfigFromFile(config);

			if (IsOptionInValid(config, args, true))
			{
				PrintUsage();
				return;
			}

			try
			{
				// Output GMSEC API version
				Log.Info(Connection.GetAPIVersion());

				// Create the connection
				conn = Connection.Create(config);

				// Connect
				conn.Connect();

				// Get CommandLine Info
				int msg_timeout_ms = Get(config, "MSG_TIMEOUT_MS", (int) GMSECWaitDefs.WAIT_FOREVER);
				int prog_timeout_s = Get(config, "PROG_TIMEOUT_S", (int) GMSECWaitDefs.WAIT_FOREVER);

				// OutPut MiddlewareVersion
				Log.Info("Middleware Version = "+conn.GetLibraryVersion());

				//Subscribe
				List<string> subjects = DetermineSubjects(config);

				for (int i = 0; i < subjects.Count; i++)
				{
					Log.Info("Subscribing to " + subjects[i]);

					info.Add(conn.Subscribe(subjects[i]));
                }

				// Wait for messages
				bool done = false;
				Stopwatch stopWatch = new Stopwatch();
				while (!done)
				{
					if (prog_timeout_s != (int)GMSECWaitDefs.WAIT_FOREVER)
					{
						if (!stopWatch.IsRunning)
						{
							stopWatch.Start();
						}
						else if (stopWatch.ElapsedMilliseconds >= (long)(prog_timeout_s * 1000))
						{
							Log.Info("Program Timeout Reached!");
							stopWatch.Stop();
							break;
						}
					}

					Message msg = conn.Receive(msg_timeout_ms);

					if (msg == null)
					{
						Log.Warning("timeout occurred");
						continue;
					}

					// Display XML represention of the received message
					Log.Info("Received Message:\n" + msg.ToXML());

					string subject = msg.GetSubject();

					if (subject.Equals("GMSEC.TERMINATE"))
					{
						done = true;
					}

                    // Check if the received message is a request message
					if (msg.GetKind() == Message.MessageKind.REQUEST)
					{
						// Construct a reply message
						string component    = msg.GetStringField("COMPONENT").GetValue();
						string statusCode   = "1";
						string replySubject = "GMSEC.MISSION.SAT_ID.RESP." + component + "." + statusCode;

						Message reply = new Message(replySubject, Message.MessageKind.REPLY);

						//Add fields to the reply message
						reply.AddField(new StringField("ANSWER", "Sure looks like it"));
						reply.AddField(new StringField("COMPONENT", "GMRPL"));

						if (config.GetBooleanValue("ENCRYPT", false))
						{
							reply.AddField(new BooleanField("ENCRYPT", true));
						}

						// Display XML representation of the reply message
						Log.Info("Prepared Reply Message:\n" + reply.ToXML());

						// Send Reply
						conn.Reply(msg, reply);
					}
				}
			}
			catch (GMSEC_Exception e)
			{
				Log.Error("Exception: " + e.ToString());
			}
			finally
			{
				CleanUp();
			}
		}


		static void Main(string[] args)
		{
			new gmrpl().Run(args);
		}
	}
}

File: examples/cpp/gmrpl/gmrpl.cpp
/*
 * Copyright 2007-2016 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */



/**
 * @file gmrpl.cpp
 * 
 * A C++ example demonstration of GMSEC request / reply functionality.
 * The associated example gmreq will provide the request.  
 * gmrpl should be run before gmreq.
 * 
 */

#include "../example.h"

#include <string>
#include <vector>

using namespace gmsec::api;


class gmrpl
{
public:
	gmrpl(Config& config);
	~gmrpl();
	bool run();

private:
	typedef std::vector<std::string> Subjects;

	Config&     config;
	Connection* connection;
	Subjects    subjects;
	SubscriptionInfo** info;

};


gmrpl::gmrpl(Config &c)
	: config(c),
	  connection(0),
	  subjects()
{
	/* Initialize config */
	example::initialize(c);
}


gmrpl::~gmrpl()
{
	if (connection)
	{
		for (size_t i = 0; i < subjects.size(); ++i)
		{
			GMSEC_INFO << "Unsubscribing from " << subjects[i].c_str();
			connection->unsubscribe(info[i]);
		}
		delete[] info;

		connection->disconnect();

		Connection::destroy(connection);
	}

	Connection::shutdownAllMiddlewares();
}


bool gmrpl::run()
{
	bool success = true;

	//o output GMSEC API version
	GMSEC_INFO << Connection::getAPIVersion();

	try
	{
		//o Create the Connection
		connection = Connection::create(config);

		//o Connect
		connection->connect();

		//o output connection middleware version
		GMSEC_INFO << connection->getLibraryVersion();

		//o Get info from command line
		std::string subject        = example::get(config, "SUBJECT", "GMSEC.MISSION.SAT_ID.REQ.CMD.GMRPL");
		int         msg_timeout_ms = example::get(config, "MSG_TIMEOUT_MS", GMSEC_WAIT_FOREVER);
		int         prog_timeout_s = example::get(config, "PROG_TIMEOUT_S", GMSEC_WAIT_FOREVER);

		//o Subscribe
		example::determineSubjects(config, subjects);

		//o Subscribe

		info = new SubscriptionInfo*[subjects.size()];
		for (size_t i = 0; i < subjects.size(); ++i)
		{
			GMSEC_INFO << "Subscribing to " << subjects[i].c_str();
			info[i] = connection->subscribe(subjects[i].c_str());
		}

		bool   done = false;
		time_t prevTime;
		time_t nextTime;
		double elapsedTime = 0;

		time(&prevTime);

		while (!done)
		{
			if (prog_timeout_s != GMSEC_WAIT_FOREVER && elapsedTime >= prog_timeout_s)
			{
				GMSEC_INFO << "Program Timeout Reached!";
				return true;
			}

			//o receive the next message
			Message* message = connection->receive(msg_timeout_ms);

			if (prog_timeout_s != GMSEC_WAIT_FOREVER)
			{
				time(&nextTime);
				elapsedTime += difftime(nextTime, prevTime);
				prevTime = nextTime;
			}

			if (!message)
			{
				GMSEC_INFO << "timeout occurred";
				continue;
			}
			else
			{
				done = (std::string(message->getSubject()) == "GMSEC.TERMINATE");

				//o Display the XML representation of the received message
				GMSEC_INFO << "Received\n" << message->toXML();

				//o check for request
				if (message->getKind() == Message::REQUEST)
				{
					const char* component = 0;

					//o Construct Reply message
					try {
						const StringField& compField = message->getStringField("COMPONENT");
						component = compField.getValue();
					}
					catch (Exception& e) {
						GMSEC_WARNING << e.what();
					}

					//o Set Status Code.  See API Interface Specification for available status codes.
					const char* status_code = "1"; // 1 = Acknowledgement

					std::ostringstream reply_subject;
					reply_subject << "GMSEC.MISSION.SAT_ID.RESP."
					              << (component ? component : "UNKNOWN-COMPONENT") << "."
					              << status_code;

					//o Create reply message
					Message reply(reply_subject.str().c_str(), Message::REPLY);

					//o Add fields to the reply message
					reply.addField("COMPONENT", "GMRPL");
					reply.addField("ANSWER", "Sure looks like it!");

					if (config.getBooleanValue("ENCRYPT", false))
					{
						reply.addField("SEC-ENCRYPT", true);
					}

					//o Display XML representation of the reply message
					GMSEC_INFO << "Prepared Reply:\n" << reply.toXML();

					//o Send Reply
					connection->reply(*message, reply);
				}

				//o Destroy request message */
				connection->release(message);
			}
		}
	}
	catch (Exception& e)
	{
		GMSEC_ERROR << e.what();
		success = false;
	}

	return success;
}


int main(int argc, char* argv[])
{
	Config config(argc, argv);

	example::addToConfigFromFile(config);

	if (example::isOptionInvalid(config, argc, "gmrpl"))
	{
		example::printUsage("gmrpl");
		return -1;
	}

	return (gmrpl(config).run() ? 0 : -1);
}
